src/kaleidoscope/hid.cpp

```diff
@@ -51,42 +52,172 @@
     pressRawKey(Key_LeftAlt);
   }
-  if (mappedKey.flags & RALT_HELD) {
+  if (flags & RALT_HELD) {
     pressRawKey(Key_RightAlt);
   }
-  if (mappedKey.flags & GUI_HELD) {
+  if (flags & GUI_HELD) {
     pressRawKey(Key_LeftGui);
   }
+}
+
+}
+
+void initializeKeyboard() {
+  Keyboard.begin();
+  WITH_BOOTKEYBOARD {
+    BootKeyboard.begin();
+  }
+}
+
+
+
+
+// pressKey takes a Key, as well as optional boolean 'toggledOn' which defaults to 'true'
+
+// If toggledOn is not set to false, this routine adds the modifier flags on this key
+// to the bitmask of modifiers that are allowed to be added to the upcoming report.
+// We do this so that when we roll over from a key with a modifier flag to one
+// without it, that modifier flag won't affect the new keypress.
+
+
+// If the key we're processing is a modifier key, any modifier flags attached to it are added
+// directly to the report along with the modifier from its keycode byte.
+//
+// (A 'modifier key' is one of the eight modifier keys defined by the HID standard:
+// left and right variants of Control, Shift, Alt, and GUI.)
+
+// Eventually it calls pressRawKey.
+
+void pressKey(Key mappedKey, boolean toggledOn) {
+  if (toggledOn) {
+    // If two keys are toggled on during the same USB report, we would ideally send an
+    // extra USB report to help the host handle each key correctly, but this is problematic.
+
+
+    // If we simply allow modifiers associated with the second newly-pressed key,
+    // it is possible to drop a modifier before the report is sent.  Instead, we send modifiers
+    // associated with any newly-pressed keys.
+
+    // The downside of this behavior is that in cases where the user presses down keys
+    // with conflicting modifiers at the exact same moment, they may get unexpected behavior.
+
+    // If this is the first 'new' keycode being pressed in this cycle, reset the bitmask
+    // of modifiers we're willing to attach to USB HID keyboard reports
+    if (!last_keycode_toggled_on) {
+      modifier_flag_mask = 0;
+    }
+
+    // Add any modifiers attached to this key to the bitmask of modifiers we're willing
+    // to attach to USB HID keyboard reports
+    modifier_flag_mask |= mappedKey.flags;
+
+    last_keycode_toggled_on = mappedKey.keyCode;
+  }
+
+
+  // If the key has any additional modifiers attached to it as flags (as one might when creating a
+  // 'Hyper' key or a "Control Alt" key, we assume that all those modifiers are intended to
+  // modify other keys pressed while this key is held, so they are never masked out.
+  if (isModifierKey(mappedKey)) {
+    pressModifiers(mappedKey.flags);
+  }
+
+  // If, instead, the modifiers are attached to a 'printable' or non-modifier key, we assume that they're
+  // not intended to modify other keys, so we add them to requested_modifiers, and only allow them to
+  // affect the report if the most recent keypress includes those modifiers.
+  else {
+    requestModifiers(mappedKey.flags);
+  }
```

Comment:
I find the big gap between the `else` and its preceding `if` block unnecessarily difficult to notice, but the code all looks correct to me.

Reply:
It's not unreasonable to move the comment inside the block. Thanks.